onnxruntime/python/tools/quantization/save_activations.py

```diff
@@ -0,0 +1,153 @@
+# --------------------------------------------------------------------------
+# Copyright (c) Microsoft, Intel Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+"""Utilities to run a given ONNX model, while saving input/output tensors of
+eligible operator nodes.
+
+A use case is to debug quantization induced accuracy drop. An AI engineer can
+run the original float32 model and the quantized model with the same inputs,
+then compare the corresponding activations between the two models to find
+where the divergence is. 
+
+Example Usage:
+
+```python
+    class ExampleDataReader(CalibrationDataReader):
+        def __init__(self):
+            ...
+        def get_next(self):
+            ...
+
+    input_data_reader = ExampleDataReader()
+
+    aug_model = augment_model_save_tensors(path_to_onnx_model)
+    augmented_model_path = str(Path(self._tmp_model_dir.name).joinpath("augmented_model.onnx"))
+    onnx.save(
+        aug_model,
+        augmented_model_path,
+        save_as_external_data=False,
+    )
+
+    tensor_dict = run_collect_activations(augmented_model_path, data_reader)
+```
+
+`tensor_dict` points to a dictionary where the keys are tensor names and each value
+is a list of tensors, one from each model run
+
+"""
+
+import time
+from pathlib import Path
+from typing import List, Optional, Union
+
+import numpy
+import onnx
+from onnx import ModelProto, TensorProto, helper, numpy_helper
+
+import onnxruntime
+
+from .calibrate import CalibraterBase, CalibrationDataReader
+from .quant_utils import clone_model_with_shape_infer
+
+_TENSOR_SAVE_POSTFIX = "_ReshapedSavedOutput"
+_TENSOR_SAVE_POSTFIX_LEN = len(_TENSOR_SAVE_POSTFIX)
+
+
+def augment_model_save_tensors(
+    onnx_model: Union[str, Path, ModelProto], op_types_for_saving: Optional[List[str]] = None
+) -> ModelProto:
+    """Augment a given ONNX model to save node input/output tensors
+
+    Add all input/output tensors of eligible operator nodes to model outputs
+    so that their value can be retrieved for debugging purposes
+
+    Args:
+        model: an ONNX model or the path to load the model
+        op_types_for_saving: Optional list of operator types for which the
+            input/output should be saved. By default, saving all the
+            float32/float16 tensors.
+
+    Returns
+        Augmented ONNX model
+    """
+
+    if op_types_for_saving is None:
+        op_types_for_saving = []
+    saver = CalibraterBase(onnx_model, op_types_to_calibrate=op_types_for_saving)
+    model = clone_model_with_shape_infer(saver.model)  # type: ModelProto
+    tensors, _ = saver.select_tensors_to_calibrate(model)
+    reshape_shape_name = "LinearReshape_" + str(time.time())
```

Comment:
Wondering if randomness is necessary here? Would a carefully chosen name works so the output is deterministic and repeatable?

Reply:
The output is a temporary model where deterministic/repeatable is not required. We are paranoid that a fixed name would eventually create a conflict with some client model, randomness gives us a way out.